packages/uploadthing/src/sdk/utils.ts

```diff
@@ -0,0 +1,86 @@
+import type { Json } from "@uploadthing/shared";
+import { generateUploadThingURL, pollForFileData } from "@uploadthing/shared";
+
+export type FileEsque = Blob & { name: string };
+
+export const uploadFilesInternal = async (
+  data: {
+    files: FileEsque[];
+    metadata: Json;
+  },
+  opts: {
+    apiKey: string;
+    utVersion: string;
+  },
+) => {
+  // Request presigned URLs for each file
+  const fileData = data.files.map((file) => ({
+    name: file.name,
+    type: file.type,
+    size: file.size,
+  }));
+  const res = await fetch(generateUploadThingURL("/api/uploadFiles"), {
+    method: "POST",
+    headers: {
+      "x-uploadthing-api-key": opts.apiKey,
+      "x-uploadthing-version": opts.utVersion,
+    },
+    cache: "no-store",
+    body: JSON.stringify({
+      files: fileData,
+      metadata: data.metadata,
+    }),
+  });
+
+  const json = (await res.json()) as
+    | {
+        data: {
+          presignedUrl: string; // url to post to
+          fields: Record<string, string>;
+          key: string;
+          fileUrl: string; // the final url of the file after upload
+        }[];
+      }
+    | { error: string };
+
+  if (!res.ok || "error" in json) {
+    const message = "error" in json ? json.error : "Unknown error";
+    throw new Error(message);
+  }
+
+  // Upload each file to S3
+  const uploads = await Promise.all(
```

Comment:
allSettled might be better here and we can handle the error cases w/ a more elegant return maybe?